source/validate_builtins.cpp

```diff
@@ -0,0 +1,1474 @@
+      assert(func_inst);
+      // The entry function uses link to the entry points
+      for (auto pair : func_inst->uses()) {
+        auto entry_point = pair.first;
+        if (entry_point->opcode() == SpvOpEntryPoint) {
+          // The entry points contain the execution model
+          auto exec_model = entry_point->words()[1];
+          exec_vec.push_back(exec_model);
+        }
+      }
+    }
+  }
+  return exec_vec;
+}
+
+// Ensures the excution model is included in the list of compatible models
+spv_result_t checkExecutionModel(ValidationState_t& vstate,
+      uint32_t exec_model, std::vector<uint32_t> models) {
+  if (not isIncludedIn(exec_model,models)) {
+    return vstate.diag(SPV_ERROR_INVALID_ID)
+        << "Built-in variables are restricted to only certain"
+           "EXECUTION MODELS (see e.g. Vulkan specification)";
+  }
+  return SPV_SUCCESS;
+}
+
+// Ensures the pair 'exec_model','storage_class' are compatible
+spv_result_t checkStorageClass(ValidationState_t& vstate,
+      uint32_t exec_model, uint32_t storage_class,
+      std::vector<uint32_t> models, std::vector<uint32_t> storages)
+{
+  // If 'exec_model' is included the list of 'models'
+  if (isIncludedIn(exec_model,models)) {
+    // But 'storage_class' is not one of the compatible 'storages'
+    if (not isIncludedIn(storage_class,storages)) {
+      return vstate.diag(SPV_ERROR_INVALID_ID)
+          << "Built-in variables must match the STORAGE CLASS "
+             "of the built-in (see e.g. Vulkan specification)";
+    }
+  }
+  return SPV_SUCCESS;
+}
+
+// Ensures the variable type matches the built-in type
+spv_result_t checkBuiltInType(ValidationState_t& vstate, Instruction inst,
+      SpvOp elem_type, SpvOp comp_type=SpvOpNop, uint32_t comp_size=0)
+{
+  auto error = [&](ValidationState_t& state) -> spv_result_t {
+    return state.diag(SPV_ERROR_INVALID_ID)
+        << "Built-in variables must match the DATA TYPE "
+           "of the built-in (see e.g. Vulkan specification)";
+  };
+  auto *curr_inst = &inst;
+  // Gets instruction type
+  curr_inst = vstate.FindDef(curr_inst->type_id());
+  assert(curr_inst);
+  // The built-in variable must be of OpType Pointer
+  if (curr_inst->opcode() != SpvOpTypePointer) {
+    return error(vstate);
+  }
+  // The Pointer type is stored in its 3rd word
+  curr_inst = vstate.FindDef(curr_inst->words()[3]);
+  assert(curr_inst);
+  // If the type is composed (i.e. Vector or Array)
+  if (comp_type != SpvOpNop) {
+    // It must match the composed type...
+    if (curr_inst->opcode() != comp_type) {
+      return error(vstate);
+    }
+    // ... and it must match the number of components / size
+    uint32_t size;
+    if (comp_type == SpvOpTypeVector) { // size as literal
+      size = curr_inst->words()[3];
+    } else if (comp_type == SpvOpTypeArray) { // size as constant
+      auto *size_inst = vstate.FindDef(curr_inst->words()[3]);
+      assert(size_inst);
+      size = size_inst->words()[3];
```

Comment:
Can the size be defined with OpSpecConstant?

Reply:
Yes, and it should work since both OpConstant and OpSpecConstant hold the literal in the word 3. BTW, shall I check whether the instruction is one of these two, or such validation logic belongs somewhere else?